Tests/UnitTests/Core/Propagator/JacobianEngineTests.cpp

```diff
@@ -0,0 +1,318 @@
+// This file is part of the Acts project.
+//
+// Copyright (C) 2020 CERN for the benefit of the Acts project
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+#include <boost/test/unit_test.hpp>
+
+#include "Acts/Definitions/Algebra.hpp"
+#include "Acts/Definitions/TrackParametrization.hpp"
+#include "Acts/Propagator/detail/JacobianEngine.hpp"
+#include "Acts/Surfaces/PlaneSurface.hpp"
+#include "Acts/Tests/CommonHelpers/FloatComparisons.hpp"
+#include "Acts/Utilities/Helpers.hpp"
+
+namespace tt = boost::test_tools;
+
+namespace Acts {
+namespace Test {
+
+/// Helper function tests
+BOOST_AUTO_TEST_CASE(jacobian_engine_helper) {
+  // (1a) Standard test with curvilinear not glazingly close to z axis
+  Vector3 direction = Vector3(7., 8., 9.).normalized();
+  FreeToBoundMatrix f2cJacobian = detail::freeToCurvilinearJacobian(direction);
+
+  ActsScalar phi = VectorHelpers::phi(direction);
+  ActsScalar theta = VectorHelpers::theta(direction);
+  ActsScalar sinPhi = std::sin(phi);
+  ActsScalar cosPhi = std::cos(phi);
+  ActsScalar sinTheta = std::sin(theta);
+  ActsScalar cosTheta = std::cos(theta);
+
+  CHECK_CLOSE_REL(f2cJacobian(eBoundLoc0, eFreePos0), -sinPhi, 1e-5);
+  CHECK_CLOSE_REL(f2cJacobian(eBoundLoc0, eFreePos1), cosPhi, 1e-5);
+  CHECK_CLOSE_REL(f2cJacobian(eBoundLoc1, eFreePos0), -cosPhi * cosTheta, 1e-5);
+  CHECK_CLOSE_REL(f2cJacobian(eBoundLoc1, eFreePos1), -sinPhi * cosTheta, 1e-5);
+  CHECK_CLOSE_REL(f2cJacobian(eBoundLoc1, eFreePos2), sinTheta, 1e-5);
+  CHECK_CLOSE_REL(f2cJacobian(eBoundTime, eFreeTime), 1., 1e-5);
+  CHECK_CLOSE_REL(f2cJacobian(eBoundPhi, eFreeDir0), -sinPhi / sinTheta, 1e-5);
+  CHECK_CLOSE_REL(f2cJacobian(eBoundPhi, eFreeDir1), cosPhi / sinTheta, 1e-5);
+  CHECK_CLOSE_REL(f2cJacobian(eBoundTheta, eFreeDir0), cosPhi * cosTheta, 1e-5);
+  CHECK_CLOSE_REL(f2cJacobian(eBoundTheta, eFreeDir1), sinPhi * cosTheta, 1e-5);
+  CHECK_CLOSE_REL(f2cJacobian(eBoundTheta, eFreeDir2), -sinTheta, 1e-5);
+  CHECK_CLOSE_REL(f2cJacobian(eBoundQOverP, eFreeQOverP), 1., 1e-5);
+
+  // (1b) Test with curvilinear glazingly close to z axis
```

Comment:
This could be its own `BOOST_AUTO_TEST_CASE`

Reply:
Further possible cases are below

Reply:
The idea here is that this tests the start Jacobian constructions, the other tests the actual transport Jacobian construction.